dags/dag_maintenance.py

```diff
@@ -0,0 +1,188 @@
+    'start_date': START_DATE,
+    'retries': 1,
+    'retry_delay': timedelta(minutes=1)
+}
+
+MAINTENANCE_DAG = DAG(
+    DAG_ID,
+    default_args=DEFAULT_ARGS,
+    schedule_interval=SCHEDULE_INTERVAL,
+    start_date=START_DATE
+)
+
+
+DEFAULT_MAX_DATA_AGE_IN_DAYS = "30"
+MAX_CLEANUP_DATA_AGE_NAME = (
+    "MAX_CLEANUP_DATA_AGE_IN_DAYS"
+)
+
+
+def get_max_data_cleanup_configuration_function(**context):
+    max_data_age_in_days = int(
+        Variable.get(
+            MAX_CLEANUP_DATA_AGE_NAME,
+            os.getenv(
+                MAX_CLEANUP_DATA_AGE_NAME,
+                DEFAULT_MAX_DATA_AGE_IN_DAYS
+            )
+        )
+    )
+    max_date = timezone.utcnow() + timedelta(-max_data_age_in_days)
+    context["ti"].xcom_push(key="max_date", value=max_date.isoformat())
+
+
+GET_CONFIGURATION = PythonOperator(
+    task_id='get_configuration',
+    python_callable=get_max_data_cleanup_configuration_function,
+    provide_context=True,
+    dag=MAINTENANCE_DAG)
+
+
+def cleanup_function(**context):
+
+    logging.info("Retrieving max_execution_date from XCom")
+    max_date = context["ti"].xcom_pull(
+        task_ids=GET_CONFIGURATION.task_id, key="max_date"
+    )
+    max_date = dateutil.parser.parse(max_date)  # stored as iso8601 str in xcom
+
+    airflow_db_model = context["params"].get("airflow_db_model")
+    age_check_column = context["params"].get("age_check_column")
+    keep_last = context["params"].get("keep_last")
+    keep_last_filters = context["params"].get("keep_last_filters")
+    keep_last_group_by = context["params"].get("keep_last_group_by")
+
+    logging.info("Running Cleanup Process...")
+    query = SESSION.query(airflow_db_model).options(
+        load_only(age_check_column)
+    )
+    if keep_last:
+        subquery = SESSION.query(func.max(DagRun.execution_date))
+        if keep_last_filters is not None:
+            for entry in keep_last_filters:
+                subquery = subquery.filter(entry)
+
+        if keep_last_group_by is not None:
+            subquery = subquery.group_by(keep_last_group_by)
+
+        subquery = subquery.from_self()
+
+        query = query.filter(
+            and_(age_check_column.notin_(subquery)),
+            and_(age_check_column <= max_date)
+        )
+
+    else:
+        query = query.filter(age_check_column <= max_date,)
+
+    if ENABLE_DELETE:
+        logging.info("Performing Delete...")
+        # using bulk delete
+        query.delete(synchronize_session=False)
+        SESSION.commit()
+        logging.info("Finished Performing Delete")
+
+    logging.info("Finished Running Cleanup Process")
+
+
+for db_object in DATABASE_OBJECTS:
+
+    cleanup_op = PythonOperator(
+        task_id='cleanup_' + str(db_object["airflow_db_model"].__name__),
+        python_callable=cleanup_function,
+        params=db_object,
+        provide_context=True,
+        dag=MAINTENANCE_DAG
+    )
+
+    GET_CONFIGURATION.set_downstream(cleanup_op)
```

Comment:
BTW from https://airflow.apache.org/docs/stable/concepts.html

> We recommend you setting operator relationships with bitshift operators rather than `set_upstream()` and `set_downstream()`.
> 
> Traditionally, operator relationships are set with the `set_upstream()` and `set_downstream()` methods. In Airflow 1.8, this can be done with the Python bitshift operators `>>` and `<<`.

I find `set_downstream` less intuitive, because if it is a regular `set` operation, then we would replace the previously set operator. `set_upstream` would avoid that problem. But maybe using a list is even better.